tests/Unit/NumericalAlgorithms/SphericalHarmonics/IO/Test_ReadSurfaceYlm.cpp

```diff
@@ -0,0 +1,358 @@
+                      std::vector<std::vector<double>> bad_data = good_data;
+                      bad_data[0][l_max_column_number] = 1.2;
+
+                      write_error_file_and_try_to_read<Frame::Inertial>(
+                          filename, "NonIntegralLmax", good_legend, bad_data,
+                          3);
+                    }()),
+                    Catch::Matchers::ContainsSubstring(
+                        "Row 0 of the Ylm data has an invalid Lmax value"));
+
+  CHECK_THROWS_WITH(([&filename, &good_legend, &good_data]() {
+                      std::vector<std::vector<double>> bad_data = good_data;
+                      bad_data[1][l_max_column_number] = -1.0;
+
+                      write_error_file_and_try_to_read<Frame::Inertial>(
+                          filename, "NegativeLmax", good_legend, bad_data, 3);
+                    }()),
+                    Catch::Matchers::ContainsSubstring(
+                        "Row 1 of the Ylm data has an invalid Lmax value"));
+
+  CHECK_THROWS_WITH(([&filename, &good_legend, &good_data]() {
+                      std::vector<std::vector<double>> bad_data = good_data;
+                      // set an Lmax that requires more columns of data than
+                      // given, i.e. `good_legend` only has room for coefs for
+                      // l <= 3
+                      bad_data[2][l_max_column_number] = 4;
+
+                      write_error_file_and_try_to_read<Frame::Inertial>(
+                          filename, "NotEnoughColumnsForLmax", good_legend,
+                          bad_data, 3);
+                    }()),
+                    Catch::Matchers::ContainsSubstring(
+                        "Row 2 of the Ylm data does not have enough "
+                        "coefficients for the Lmax. For Lmax = 4, expected at "
+                        "least 25 coefficient columns and 30 total columns"));
+
+  CHECK_THROWS_WITH(
+      ([&filename, &good_legend, &good_data]() {
+        std::vector<std::vector<double>> bad_data = good_data;
+        bad_data[0].back() = 0.8;
+
+        write_error_file_and_try_to_read<Frame::Inertial>(
+            filename, "NonZeroHigherCoefs", good_legend, bad_data, 3);
+      }()),
+      Catch::Matchers::ContainsSubstring(
+          "Row 0 of the Ylm data has Lmax = 2 but non-zero coefficients for "
+          "l > Lmax"));
+  if (file_system::check_if_file_exists(filename)) {
+    file_system::rm(filename, true);
+  }
+}
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.SphericalHarmonics.ReadSurfaceYlm",
+                  "[NumericalAlgorithms][Unit]") {
+  test_errors();
+
+  // Temporary file with test data to read in
+  const std::string test_filename{"TestReadYlm.h5"};
+  if (file_system::check_if_file_exists(test_filename)) {
+    file_system::rm(test_filename, true);
+  }
+
+  // first test surface
+  using frame_a = Frame::Grid;
+  const std::string subfile_name_a = "SurfaceA_Ylm";
```

Comment:
(optional, you can do this while squashing) You can avoid the `_a` etc suffixes by using scopes:

```cpp
{
  INFO("Test A");
  using frame = Frame::Grid;
  ...
}
{
  INFO("Test B");
  // Here you can use the same variable names
}
```

This also gives better test output because you can use `INFO` and `CAPTURE` limited to the scope.